frontend/src/poc-code/testK8s/ws/WSFactory.ts

```diff
@@ -0,0 +1,338 @@
+      console.log(`websocket error: ${this.id}`);
+      this.state = 'error';
+      this.triggerEvent('error', evt);
+    };
+    this.ws.onmessage = (evt) => {
+      const msg = this.options?.jsonParse ? JSON.parse(evt.data) : evt.data;
+      // In some browsers, onmessage can fire after onclose/error. Don't update state to be incorrect.
+      if (this.state !== 'destroyed' && this.state !== 'closed') {
+        this.state = 'open';
+      }
+      this.triggerEvent('message', msg);
+    };
+  }
+
+  /**
+   * Invoke all registered handler callbacks for a given data.
+   */
+  private invokeHandlers(type: EventHandlerTypes, data?: unknown) {
+    const handlers = this.handlers[type];
+    if (!handlers) {
+      return;
+    }
+    handlers.forEach((h) => {
+      try {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        h(data); // typescript is having an issue with passing the data, muting for now
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error('WS handling failed:', e);
+      }
+    });
+  }
+
+  /**
+   * Triggers event to be buffered or invoked depending on config.
+   */
+  private triggerEvent(type: EventHandlerTypes, data?: unknown) {
+    if (this.state === 'destroyed') {
+      return;
+    }
+
+    const isMessageEvent = (t: string, e: unknown): e is MessageDataType => {
+      return t === 'message' && !!e;
+    };
+
+    // Only buffer "message" events, so "error" and "close" etc can pass thru.
+    if (this.bufferMax && isMessageEvent(type, data)) {
+      this.messageBuffer.push(data);
+
+      if (this.messageBuffer.length > this.bufferMax) {
+        this.messageBuffer.shift();
+      }
+
+      return;
+    }
+
+    this.invokeHandlers(type, data);
+  }
+
+  onMessage(fn: MessageHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.message.push(fn);
+    return this;
+  }
+
+  onBulkMessage(fn: BulkMessageHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.bulkMessage.push(fn);
+    return this;
+  }
+
+  onError(fn: ErrorHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.error.push(fn);
+    return this;
+  }
+
+  onOpen(fn: OpenHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.open.push(fn);
+    return this;
+  }
+
+  onClose(fn: CloseHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.close.push(fn);
+    return this;
+  }
+
+  onDestroy(fn: DestroyHandler): WSFactory {
+    if (this.state === 'destroyed') {
+      return this;
+    }
+    this.handlers.destroy.push(fn);
+    return this;
+  }
+
+  flushMessageBuffer() {
+    if (this.paused) {
+      return;
+    }
+
+    if (!this.messageBuffer.length) {
+      return;
+    }
```

Comment:
Can we merge these two ifs together?

Reply:
Sure, why not. Will be in https://github.com/openshift/dynamic-plugin-sdk/pull/38